Add upload manager

akosel opened the pull request:
Adds upload manager service that keeps track of uploads currently in progress.

zombiezen commented on client/uploadmanager.js:
```diff
@@ -0,0 +1,129 @@
+ * limitations under the License.
+ */
+
+//goog.provide('filedrop.Files');
```
`filedrop.UploadManager`, please.

akosel committed:
Add upload manager

akosel committed:
Merge branch 'master' into add-upload-manager

zombiezen commented:
I'll try to review this over the weekend.

zombiezen commented:
Sorry, that weekend lasted several months.

zombiezen commented on client/uploadmanager.js:
```diff
@@ -0,0 +1,129 @@
+ */
+filedrop.UploadManager.prototype.add = function(file) {
+  var upload = this.filesService_.upload(file);
+  var id = file.name;
```
Make this an incrementing counter instead so as to avoid overwriting existing uploads.

zombiezen commented on client/uploadmanager.js:
```diff
@@ -0,0 +1,129 @@
+ * @param {!File} file
+ * @return {!angular.$q.Promise}
+ */
+filedrop.UploadManager.prototype.makeUpload = function(file) {
```
Can this be made the same function as add?

zombiezen commented on client/uploadmanager.js:
```diff
@@ -0,0 +1,129 @@
+filedrop.UploadManager.prototype.fetch = function(ids) {
+  var promises = [];
+
+  if (!ids) {
```
Roll this logic into fetchAll; don't clutter fetch with it.

zombiezen commented on client/uploadmanager.js:
```diff
@@ -0,0 +1,129 @@
+  }
+
+  return this.q_.all(promises).then(angular.bind(this, function(val) {
+    return arguments[0];
```
Isn't this just the identity function?

zombiezen commented on client/controller.js:
```diff
@@ -136,14 +139,21 @@ filedrop.Controller.prototype.dropFile = function(files) {
+        }, this);
+        this.toast_.showSimple(okText);
+      }), angular.bind(this, function() {
+        var failText = 'Failed to upload a file';
```
It would be nice to show a simple count of how many succeeded and how many failed, but maybe this is OK until we add a UI for this manager.

akosel replied:
Yeah, I started on the UI portion. I can add a way to keep track of failures/successes in that PR.

zombiezen commented:
Overall, LGTM. Thanks again for this.

akosel commented:
Sounds like it must have been a good weekend! :)

Thanks for all of the comments; I'll fix this up in the next week or so.

akosel committed:
Simplify fetch. Use auto-incrementing id for files.

akosel committed:
Remove makeUpload, as it is unused right now

akosel commented:
Yeah, that week lasted a little long too. Anyways, I believe I addressed all the comments. I decided to remove `makeUpload` since it isn't really needed, at least not at the moment.

zombiezen commented:
:+1:
